rpc/src/v1/types/block_number.rs

```diff
@@ -131,19 +193,40 @@ pub fn block_number_to_id(number: BlockNumber) -> BlockId {
 mod tests {
 	use ethcore::client::BlockId;
 	use super::*;
+	use std::str::FromStr;
 	use serde_json;
 
 	#[test]
 	fn block_number_deserialization() {
-		let s = r#"["0xa", "latest", "earliest", "pending"]"#;
+		let s = r#"[
+			"0xa",
+			"latest",
+			"earliest",
+			"pending",
+			{"blockNumber": "0xa"},
+			{"blockHash": "0x1dcc4de8dec75d7aab85b567b6ccd41ad312451b948a7413f0a142fd40d49347"},
+			{"blockHash": "0x1dcc4de8dec75d7aab85b567b6ccd41ad312451b948a7413f0a142fd40d49347", "requireCanonical": true}
+		]"#;
 		let deserialized: Vec<BlockNumber> = serde_json::from_str(s).unwrap();
-		assert_eq!(deserialized, vec![BlockNumber::Num(10), BlockNumber::Latest, BlockNumber::Earliest, BlockNumber::Pending])
+
+		assert_eq!(
+			deserialized,
+			vec![
+				BlockNumber::Num(10),
+				BlockNumber::Latest,
+				BlockNumber::Earliest,
+				BlockNumber::Pending,
+				BlockNumber::Num(10),
+				BlockNumber::Hash { hash: H256::from_str("1dcc4de8dec75d7aab85b567b6ccd41ad312451b948a7413f0a142fd40d49347").unwrap(), require_canonical: false },
+				BlockNumber::Hash { hash: H256::from_str("1dcc4de8dec75d7aab85b567b6ccd41ad312451b948a7413f0a142fd40d49347").unwrap(), require_canonical: true }
+			]
+		)
 	}
 
 	#[test]
-	fn should_not_deserialize_decimal() {
-		let s = r#""10""#;
-		assert!(serde_json::from_str::<BlockNumber>(s).is_err());
+	fn should_not_deserialize() {
+		let s = r#"[{}, "10"]"#;
```

Comment:
Do we reject `requireCanonical` with `BlockNumber` or any other invalid combinations?

Reply:
Currently, the code does not reject if used with `blockNumber` instead it is ignored. Do you think we should reject?

Reply:
I think if derive `Serialize/Deserialize` (seems possible with serde attributes) instead of manually implementing it, it would be automatically supported. This can be done in a separate PR though.